python/ray/monitor.py

```diff
@@ -478,6 +482,104 @@ def driver_removed_handler(self, unused_channel, data):
 
         self._clean_up_entries_for_driver(driver_id)
 
+    def _xray_clean_up_entries_for_driver(self, driver_id):
+        """Remove this driver's object/task entries from redis.
+
+        Removes control-state entries of all tasks and task return
+        objects belonging to the driver.
+
+        Args:
+            driver_id: The driver id.
+        """
+        redis = self.state.redis_clients[0]
+
+        xray_task_table_prefix = (
+            ray.gcs_utils.TablePrefix_RAYLET_TASK_string.encode("ascii"))
+        xray_object_table_prefix = (
+            ray.gcs_utils.TablePrefix_OBJECT_string.encode("ascii"))
+
+        task_table_infos = {}  # task id -> TaskInfo
+        for key in redis.scan_iter(match=xray_task_table_prefix + b"*"):
+            entry = redis.get(key)
+            if entry is None:
+                continue
+            task = ray.gcs_utils.Task.GetRootAsTask(entry, 0)
+            task_info = ray.gcs_utils.TaskInfo.GetRootAsTaskInfo(
+                task.TaskSpecification(), 0)
+            if driver_id != task_info.DriverId():
+                # Ignore tasks that aren't from this driver.
+                continue
+            task_table_infos[task_info.TaskId()] = task_info
+        driver_task_ids = task_table_infos.keys()
+
+        # Get the list of objects returned by driver tasks.
+        driver_object_ids = []
+        for task_info in task_table_infos.values():
+            driver_object_ids.extend([
+                task_info.Returns(i) for i in range(task_info.ReturnsLength())
+            ])
+
+        # Also record all the ray.put()'d objects.
+        all_put_objects = []
+        for key in redis.scan_iter(match=xray_object_table_prefix + b"*"):
+            entry = redis.zrange(key, 0, 0)
+            if entry is None:
+                continue
+            assert len(entry) == 1
```

Comment:
I suspect this `assert` will be False a lot of the time (e.g., for an object that was transferred to a different machine or was evicted).

Reply:
Oh, I see you're using `zrange(key, 0, 0)`. Maybe instead we should use `RAY.TABLE_LOOKUP`.

Reply:
I think this type of refactoring makes sense to apply to the entire module in a subsequent PR, so that the module implementation remains consistent.

Reply:
I replaced this loop with a call to `self.state.object_table()`.